src/Catalyst.Core.Lib/FileSystem/FileStore.cs

```diff
@@ -0,0 +1,348 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Nito.AsyncEx;
+
+namespace Catalyst.Core.Lib.FileSystem
+{
+    /// <summary>
+    ///     A file based repository for name value pairs.
+    /// </summary>
+    /// <typeparam name="TName">
+    ///     The type used for a unique name.
+    /// </typeparam>
+    /// <typeparam name="TValue">
+    ///     The type used for the value.
+    /// </typeparam>
+    /// <remarks>
+    ///     All operations are atomic, a reader/writer lock is used.
+    /// </remarks>
+    public class FileStore<TName, TValue>
+        where TValue : class
+    {
+        private readonly AsyncReaderWriterLock _storeLock = new AsyncReaderWriterLock();
+
+        /// <summary>
+        ///     A function to write the JSON encoded entity to the stream.
+        /// </summary>
+        /// <remarks>
+        ///     This is the default <see cref="Serialize" />.
+        /// </remarks>
+        public static Func<Stream, TName, TValue, CancellationToken, Task> JsonSerialize =
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)